### ISSUE TITLE
Updating KDE Container

### ISSUE THREAD

**Issue by adriaandegroot** [JanitorTechnology/janitor#300]
The KDE container was created by rca, but it hasn't been updated recently, and could probably use some refinement. Various questions:

 - one container / image, or multiple containers / images for different working areas? I can imagine an application developer has different (dev) needs than a frameworks developer, and different bundles of source-packages installed.
 - what base to use? There's ubuntu-dev (which is apparently used by other images as well), but Neon on Krypton might make sense as well or instead -- it depends a little on what kind of developer we're trying to cater to.

**Comment by bundito** [comment]
I would cast my two-cent vote for Neon, as it's an existing KDE product/project which this would help promote. That said, I can't comment on any of the technical differences, just the additional exposure Neon would get from this. Plus, as a KDE newcomer and Neon Dev Unstable user, there's a certain amount of "pride" that I get from having access to developer resources - things that seem unobtainable in many other circumstances.

**Comment by jankeromnes** [comment]
Hi @adriaandegroot and @bundito, thanks a lot for opening this issue!

For reference, our [current KDE image](https://janitor.technology/projects/) on Janitor is indeed based on @rcatolino's Dockerfile: https://github.com/rcatolino/kdesrcbuild-docker/ which is published on Docker Hub as: https://hub.docker.com/r/janitortechnology/kde/

It's now slightly outdated, and should probably be rebased on top of [ubuntu-dev](https://github.com/JanitorTechnology/dockerfiles/blob/master/ubuntu-dev/ubuntu-dev.dockerfile) just like most of our [dockerfiles](https://github.com/JanitorTechnology/dockerfiles) (for an example, see the short [Rust Dockerfile](https://github.com/JanitorTechnology/dockerfiles/blob/master/rust/rust.dockerfile) that was recently added to Janitor). `ubuntu-dev` is what's most compatible with Janitor today: standard `ubuntu:16.04` with many useful dev tools like `git`, `clang`, `vim`... in their latest stable versions, a web-based VNC, web-based IDEs, etc. (That being said, it's possible to use a different base image, but the Janitor service expects a project workspace that lives in `/home/user/$PROJECT`; a web terminal server running on port 8087; a noVNC server on port 8088; a Cloud9 IDE on port 8089; and since recently a Theia IDE on port 8090.)

> - one container / image, or multiple containers / images for different working areas?

Large images aren't a problem for Janitor (in fact, it's even better to include all tools that developers would ever need into a single image, because Copy-on-Write guarantees that these tools will be saved once on our server's disk, not `#developers * #containers` times).

But then it depends on how different the workflows are, e.g. if KDE app developers don't use the same tools or workflows than KDE libs developers or KDE desktop shell developers at all, then we might as well make 3 separate images.

We just need to know the commands to set up a development environment; and a few useful one-liners typically used by developers. Schematically, current Janitor project images are usually built like this:

```bash
# project.dockerfile: A development environment based on Ubuntu.
FROM janx/ubuntu-dev
RUN apt-get install dependencies
RUN git clone https://github.com/project/ /home/user/project
COPY --chown=user janitor.json /home/user/
EXPOSE 8080
```

```js
// janitor.json: A configuration file describing the project, any exposed web ports, and a few useful one-liners.
{
  "name": "Project",
  "ports": {
    "8080": {
      "label": "Optional Web Preview",
      "proxy": "https"
    }
  },
  "scripts": {
    "Build Project": "./build",
    "Run Project": "./start --option",
    "Test Project": "./tests/run.sh",
    "Update source code": "git pull --rebase origin master",
    "Send to code review": "hub pull-request"
  }
}
```

We'd love to refresh our KDE image, and make it useful to all KDE developers! 😄

**Comment by rcatolino** [comment]
> We just need to know the commands to set up a development environment;

For what it's worth I've built the existing container by following [https://community.kde.org/Guidelines_and_HOWTOs/Build_from_source](https://community.kde.org/Guidelines_and_HOWTOs/Build_from_source) and putting all the steps in a dockerfile (more or less). It's all rather well documented.

I started rebasing it on top of ubuntu 16.04 yesterday, and it seems like the qt-dev packages from the repos are too old too build KDE. I'm using a ppa right now, but switching to neon would probably be easier to maintain in the long run since we wouldn't have to worry about this kind of compatibility issues. On the other hand we would have to replicate most of [ubuntu-dev](https://github.com/JanitorTechnology/dockerfiles/blob/master/ubuntu-dev/ubuntu-dev.dockerfile) in the dockerfile.

I think I saw somewhere that neon was also based on ubuntu 16.04, is that right ? In that case would it be possible to install neon on top of [ubuntu-dev](https://github.com/JanitorTechnology/dockerfiles/blob/master/ubuntu-dev/ubuntu-dev.dockerfile)?

**Comment by bundito** [comment]
> I think I saw somewhere that neon was also based on ubuntu 16.04, is that right ? In that case would it be possible to install neon on top of ubuntu-dev?

Yes, neon is Ubuntu 16.04. https://neon.kde.org/faq#what-is-neon

I know that much. Can't help with the ubuntu-dev part, though. Thanks for your efforts!
